sys/riscv/exception.S

```diff
@@ -1,13 +1,300 @@
 #include <riscv/asm.h>
+#include <riscv/riscvreg.h>
 
-ENTRY(cpu_exception_handler)
-	# Not implemented!
-	j .
-END(cpu_exception_handler)
+#include "assym.h"
 
-ENTRY(user_exc_leave)
-	# Not implemented!
-	j .
-END(user_exc_leave)
+.macro save_reg reg, offset
+	REG_S	\reg, (\offset)(sp)
+.endm
+
+.macro save_reg_cfi reg, offset
+	save_reg \reg, \offset
+	.cfi_rel_offset	\reg, \offset
+.endm
+
+.macro save_csr csr, offset, tmp
+	csrr	\tmp, \csr
+	save_reg \tmp, \offset
+.endm
+
+.macro load_reg reg, offset
+	REG_L	\reg, (\offset)(sp)
+.endm
+
+.macro load_csr csr, offset, tmp
+	load_reg \tmp, \offset
+	csrw	\csr, \tmp
+.endm
```

Comment:
It is used also in `switch.S`. Could you share this code, we are already using cpp for asm files so it should be simple.

Reply:
I moved common macros to `include/riscv/asm.h`.